vivisect/notes/vivisect/impapi/posix_i386_generator_gnuorg.py

```diff
@@ -0,0 +1,192 @@
+import os
+import bs4
+import pickle
+import requests
+import tempfile
+'''
+Downloads lots of pages, parses, and pulls out function declarations for impapi
+
+(210409)
+One problem: 7 functions from "Waiting with Specific Clocks" page don't have 
+all their arguments fit in the <dt> tags.  Rather than rework the script that
+works for 1500+ other functions, we just modified the output manually.
+(https://www.gnu.org/software/libc/manual/html_node/Waiting-with-Explicit-Clocks.html)
+
+
+'''
+
+BASE_URL = "https://www.gnu.org/software/libc/manual/html_node/"
+def cache_store(name, obj):
+    tmpfile = os.sep.join([tempfile.gettempdir(), ".impapi.posix."+name])
+    print(repr(tmpfile))
+    outfile = open(tmpfile, 'wb')
+    pcklstr = pickle.dumps(obj)
+    outfile.write(pcklstr)
+
+def cache_retr(name):
+    tmpfile = os.sep.join([tempfile.gettempdir(), ".impapi.posix."+name])
+    try:
+        outf = open(tmpfile, 'rb')
+        return pickle.load(outf)
+    except Exception as e:
+        print(e)
+    return None
+
+
+def getAllFrontUrls(sess):
+    cached = cache_retr('urls')
+    if cached is not None:
+        return cached
+
+    frontpage = sess.get(BASE_URL + 'Function-Index.html') 
+    soup = bs4.BeautifulSoup(frontpage.text)
+    urls = []
+    for link in soup.find_all('a'):
+        url = link.get('href') 
+        uurl = url 
+        if '#' in uurl: 
+            uurl = url[:url.find('#')] 
+ 
+        if url in urls:
+            continue
+
+        urls.append(uurl)
+
+    cache_store('urls', urls)
+
+    return urls
+
+
+def getAllFuncProtos(funcs=None):
+    if funcs is None:
+        funcs = getRawDataLines()
+    api = parseDataLines(funcs)
+
+    return api
+
+
+def getRawDataLines():
+    pages = cache_retr('pages')
+    if pages is not None:
+        print("using cached `pages`: %s" % repr(pages)[:20])
+
+    else:
+        print("no cache, pulling pages from source...")
+        sess = requests.session()
+
+        # get all 1500+ urls and downselect them to functions 
+        urls = getAllFrontUrls(sess)
+        pages = [sess.get(BASE_URL + url) for url in urls]
+
+        # store the cache in /tmp
+        cache_store('pages', pages)
+
+    bss = [bs4.BeautifulSoup(page.text) for page in pages]
+
+    dts = []
+    funcdts = []
+    for b in bss:
+        dt = b.find_all('dt')
+
+        for item in dt:
+            if 'Function:' in repr(item):
+                if item not in funcdts:
+                    funcdts.append(item)
+
+    # strip out the text and whittle each function down to a prototype line
+    functext = [fdt.text for fdt in funcdts]
+    funcs = [line[10:] for line in functext]
+    return funcs
+
+def parseDataLines(funcs):
+    # parse each line and populate the api dict
+    api = {}
+    for func in funcs:
+        func = func.replace("*", "* ").replace(';', '')
+        print(func)
```

Comment:
print -> logging

And can you make the message more than just the function name so we can determine what's going on at a glance?